go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go

```diff
@@ -608,3 +608,25 @@ func (jsIPManager *JoinSwitchIPManager) ReleaseJoinLRPIPs(nodeName string) (err
 	}
 	return err
 }
+
+// NewL2SwitchManager initializes a new layer2 logical switch manager,
+// only manage subnet for the one specified switch
+func NewL2SwitchManager() *LogicalSwitchManager {
+	return &LogicalSwitchManager{
+		cache:    make(map[string]logicalSwitchInfo),
+		RWMutex:  sync.RWMutex{},
+		ipamFunc: NewL2IPAMAllocator,
+	}
+}
+
+// NewLayer2IPAMAllocator provides an ipam interface which can be used for layer2 switch IPAM
+// allocations for the specified cidr using a contiguous allocation strategy.
+func NewL2IPAMAllocator(cidr *net.IPNet) (ipam.Interface, error) {
+	subnetRange, err := ipam.NewAllocatorCIDRRange(cidr, func(max int, rangeSpec string) (allocator.Interface, error) {
+		return allocator.NewRoundRobinAllocationMap(max, rangeSpec), nil
+	})
+	if err != nil {
+		return nil, err
+	}
```

Comment:
that would require quite some changes of the existing  logicalswitchManager interface, to carry extra exludedIPs information.

Reply:
It would only require to change `NewL2SwitchManager()`? Then the ipamFunc could be a closure which checks which of the excluded subnets overlaps with the provided CIDR (via utility function). And then the allocator itself could be updated to be able to allocate ranges or if, indeed, this specifically is too much of an effort and change , this would be the place to iterate all IPs and reserve them, I think.

On another note, having the Netconf subnets as a json array of json objects with two properties, a subnet itself and a excluded subnet that should overlap, could probably set us up better to have simpler code.

Reply:
not sure I understand your proposal, do it mean we change the signature of ipamFactoryFunc to take excluded subnets as well?

Reply:
I was thinking on something like this
```
func NewL2SwitchManager(reserveSubnets []*net.ipNet) *LogicalSwitchManager {
        
        ipamFunc := func (cidr *net.IPNet) (ipam.Interface, error) {
           if i, err := NewL2IPAMAllocator(cidr); err != nil {
              return err
           }
           for _, reservedSubnet := range reservedSubnets {
              if subnetOverlaps(cidr, reservedSubnet) {
                 err := reserveSubnet(reservedSubnet, cidr, ipamFunc)
              }
           }
        }
        
	return &LogicalSwitchManager{
		cache:    make(map[string]logicalSwitchInfo),
		RWMutex:  sync.RWMutex{},
		ipamFunc: ipamFunc,
	}
}
```

But it might make more sense to have an `AllocateSubnets` in addition to `AllocateIPs` in the LS manager itself.

Reply:
Hmm I am thinking let's not do the NewL2SwitchManager I am proposing here because now that I see it I am not convinced myself. 

Having `AllocateSubnets` could still make sense. The idea being we limit iterating IPs to the allocator itself until we handle that better.